File: CoinChangeExhaustive.java
// Time Complexity : O(2 pow (m + n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int coinChange(int[] coins, int amount) {
        return helper(coins, amount, 0, 0);
    }
    
    public int helper(int[] coins,int amount,int index, int min){
        //base case
        if(amount == 0)
            return min;
        if(amount < 0 || index > coins.length - 1)
            return -1;
        
        //do not consider the coin
        int case1 = helper(coins,amount,index + 1,min);
        
        //consider the coin
        int case2 = helper(coins,amount - coins[index],index,min + 1);
        
        if(case1 == -1)
            return case2;
        else if(case2 == -1)
            return case1;
        //return minimum of the two cases
        return Math.min(case1,case2);
    }
}

File: coin-change-exhaustive.cpp
// Time Complexity : O(2^(m+n))
// Space Complexity : O(m*n)
// Did this code successfully run on Leetcode : No since time complexity is exponential which is why for some cases time limit exceeds.

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        return helper(coins, 0, amount, 0);
    }
    int helper(vector<int>& coins, int idx, int amount, int coinsUsed){
        //base
        if(amount == 0) return coinsUsed;
        if(amount < 0 || idx == coins.size()) return -1;
        //logic
        //choose
        int chose = helper(coins, idx, (amount - coins[idx]), coinsUsed + 1);
        //dont choose
        int reject = helper(coins, idx + 1, amount, coinsUsed);

        if(chose == -1) return reject;
        if(reject == -1) return chose;

        return min(chose, reject);

    }
};